File: backend/internal/server/message.go
package server

import (
	"encoding/json"
	"go-chess/internal/auth"
	"go-chess/internal/chess"
)

type MessageType string

const (
	// outbound
	MessageTypeBoard             MessageType = "board"
	MessageTypeMatchmakingUpdate MessageType = "matchmaking_update"
	MessageTypeStartMatch        MessageType = "start_match"
	MessageTypeEndMatch          MessageType = "end_match"
	MessageTypeDrawOffered       MessageType = "draw_offered"
	MessageTypeDrawDeclined      MessageType = "draw_declined"
	MessageTypePlayerInfo        MessageType = "player_info"
	// inbound
	MessageTypeJoinMatch   MessageType = "join_match"
	MessageTypeLeaveMatch  MessageType = "leave_match"
	MessageTypeMove        MessageType = "move"
	MessageTypeResign      MessageType = "resign"
	MessageTypeOfferDraw   MessageType = "offer_draw"
	MessageTypeRespondDraw MessageType = "respond_draw"
)

type WSMessage struct {
	Type MessageType     `json:"type"`
	Data json.RawMessage `json:"data"`
}

// outbound
type LegalMove struct {
	To        string           `json:"to"`
	Promotion *chess.PieceType `json:"promotion,omitempty"`
}

type ClockData struct {
	WhiteTimeMs int64 `json:"whiteTimeMs"`
	BlackTimeMs int64 `json:"blackTimeMs"`
	IncrementMs int64 `json:"incrementMs"`
}

type BoardData struct {
	Board       chess.Board            `json:"board"`
	LegalMoves  map[string][]LegalMove `json:"legalMoves"`
	PGN         chess.PGN              `json:"pgn"`
	ActiveColor chess.Color            `json:"activeColor"`
	Clock       ClockData              `json:"clock"`
}

type TimeFormatMs struct {
	InitialMs   int64 `json:"initialMs"`
	IncrementMs int64 `json:"incrementMs"`
}

type QueueData struct {
	TimeFormat  TimeFormatMs `json:"timeFormat"`
	QueueLength int          `json:"queueLength"`
	InQueue     bool         `json:"inQueue"`
}

type MatchmakingUpdateData struct {
	Queues []QueueData `json:"queues"`
}

type StartMatchData struct {
	PublicID        string      `json:"publicId"`
	Color           chess.Color `json:"color"`
	WhitePlayerName string      `json:"whitePlayerName"`
	BlackPlayerName string      `json:"blackPlayerName"`
	Clock           ClockData   `json:"clock"`
}

type EndMatchData struct {
	Result chess.Result `json:"result"`
}

type PlayerInfoData = auth.PlayerInfoData

// inbound
type JoinMatchData struct {
	TimeFormat TimeFormatMs `json:"timeFormat"`
}

type LeaveMatchData struct {
	TimeFormat TimeFormatMs `json:"timeFormat"`
}

type MoveData struct {
	From      string           `json:"from"`
	To        string           `json:"to"`
	Promotion *chess.PieceType `json:"promotion,omitempty"`
}

type RespondDrawData struct {
	Accept bool `json:"accept"`
}

File: frontend/src/types/message.ts
import type { Board, Color, PieceType, TimeFormat } from "./chess";
import type { Result } from "./result";

// inbound
const MessageTypeBoard = "board";
const MessageTypeMatchmakingUpdate = "matchmaking_update";
const MessageTypeStartMatch = "start_match";
const MessageTypeEndMatch = "end_match";
const MessageTypeDrawOffered = "draw_offered";
const MessageTypeDrawDeclined = "draw_declined";
const MessageTypePlayerInfo = "player_info";
// outbound
const MessageTypeJoinMatch = "join_match";
const MessageTypeLeaveMatch = "leave_match";
const MessageTypeMove = "move";
const MessageTypeResign = "resign";
const MessageTypeOfferDraw = "offer_draw";
const MessageTypeRespondDraw = "respond_draw";

// inbound
type LegalMove = {
  to: string;
  promotion: PieceType | null;
};

type ClockData = {
  whiteTimeMs: number;
  blackTimeMs: number;
  incrementMs: number;
};

type BoardData = {
  board: Board;
  legalMoves: Record<string, LegalMove[]>;
  pgn: string;
  activeColor: Color;
  clock: ClockData;
};

type QueueData = {
  timeFormat: TimeFormat;
  queueLength: number;
  inQueue: boolean;
};

type MatchmakingUpdateData = {
  queues: QueueData[];
};

type StartMatchData = {
  color: Color;
  whitePlayerName: string;
  blackPlayerName: string;
  clock: ClockData;
};

type EndMatchData = {
  result: Result;
};

type PlayerInfoData = {
  displayName: string;
  username: string;
  isAuthenticated: boolean;
};

// outbound
type JoinMatchData = {
  timeFormat: TimeFormat;
};

type LeaveMatchData = {
  timeFormat: TimeFormat;
};

type MoveData = {
  from: string;
  to: string;
  promotion: PieceType | null;
};

type RespondDrawData = {
  accept: boolean;
};

type WSMessage =
  // inbound
  | {
      type: typeof MessageTypeBoard;
      data: BoardData;
    }
  | {
      type: typeof MessageTypeStartMatch;
      data: StartMatchData;
    }
  | {
      type: typeof MessageTypeMatchmakingUpdate;
      data: MatchmakingUpdateData;
    }
  | {
      type: typeof MessageTypeEndMatch;
      data: EndMatchData;
    }
  | {
      type: typeof MessageTypeDrawOffered;
    }
  | {
      type: typeof MessageTypeDrawDeclined;
    }
  | {
      type: typeof MessageTypePlayerInfo;
      data: PlayerInfoData;
    }
  // outbound
  | {
      type: typeof MessageTypeJoinMatch;
      data: JoinMatchData;
    }
  | {
      type: typeof MessageTypeLeaveMatch;
      data: LeaveMatchData;
    }
  | {
      type: typeof MessageTypeMove;
      data: MoveData;
    }
  | {
      type: typeof MessageTypeResign;
    }
  | {
      type: typeof MessageTypeOfferDraw;
    }
  | {
      type: typeof MessageTypeRespondDraw;
      data: RespondDrawData;
    };

export type {
  WSMessage,
  MoveData,
  BoardData,
  JoinMatchData,
  LegalMove,
  StartMatchData,
  MatchmakingUpdateData,
  EndMatchData,
  ClockData,
  QueueData,
  PlayerInfoData,
};
export {
  MessageTypeBoard,
  MessageTypeMove,
  MessageTypeJoinMatch,
  MessageTypeLeaveMatch,
  MessageTypeStartMatch,
  MessageTypeMatchmakingUpdate,
  MessageTypeEndMatch,
  MessageTypeDrawOffered,
  MessageTypeDrawDeclined,
  MessageTypeResign,
  MessageTypeOfferDraw,
  MessageTypeRespondDraw,
  MessageTypePlayerInfo,
};
